### PR TITLE
[android] Add support for android's file descriptor ownership tagging to libstd.

### PR DESCRIPTION
**Pull request by jmgao**
Android has started investigating using Rust for system components, and one of the biggest historical sources of bugs we've had in cross-language interop has been caused by confusion around file descriptor ownership, since you're forced to pass file descriptors across language boundaries as a bare int without any type information about whether ownership is being transferred. The bugs this results in are horrible to debug, because they're nondeterministic and manifest as impossible behavior occurring on different threads. (e.g. [[1](https://www.sqlite.org/howtocorrupt.html#_continuing_to_use_a_file_descriptor_after_it_has_been_closed)], [[2](https://engineering.fb.com/ios/debugging-file-corruption-on-ios/)])

We've added [some facilities to our libc](https://android.googlesource.com/platform/bionic/+/refs/heads/master/libc/include/android/fdsan.h) to detect/prevent this class of bugs. Basically, user code can tell libc that a file descriptor can only be closed with a specified token, and to abort/log an error if someone tries to close the fd without the token (the behavior is process-wide and configurable at runtime, but the default on Android R (API level 30) and beyond is to abort). We've found dozens of bugs in Android with this, and hopefully prevented many more from ever being checked in.

This PR makes libstd use this on android when available. It potentially changes behavior, but only for code that's either broken, or very weird (e.g. `File::from_raw_fd(fd).into_raw_fd()` sequences which construct a File that owns the fd, and then removes it from the File before drop), and also only for unsafe code, I believe.

~(also, as another commit in this PR, I improved the error handling on non-android close. Let me know if I should split this out to a separate PR)~ (whoops, didn't actually build this commit: I'll do this as a followup)

### CODE REVIEW

**Comment by nikomatsakis** [comment]
I left a comment about this on Zulip -- I'm not sure the proper procedure here, but I'm inclined to land this PR. Probably it should be tagged with `relnotes` though.

**Comment by nikomatsakis** [comment]
Thinking a bit about it, since it is a user-visible change, I think an FCP is probably appropriate. @jmgao I'm curious if you could give some examples of what kinds of bugs you expect to find with this change -- e.g., would it help to detect broken code from users? Only if they are using the `from_raw_fd` API?

**Comment by nikomatsakis** [comment]
Also, cc @rust-lang/libs -- this seems to lie somewhere on the boundary of "implementation detail" and "public facing API change", so maybe y'all want to weigh in?

**Comment by Amanieu** [comment]
Should `into_raw_fd` reset the tag to 0? I think it should since you're explicitly saying that the fd is no longer managed by Rust.

**Comment by jmgao** [comment]
> Thinking a bit about it, since it is a user-visible change, I think an FCP is probably appropriate. @jmgao I'm curious if you could give some examples of what kinds of bugs you expect to find with this change -- e.g., would it help to detect broken code from users? Only if they are using the `from_raw_fd` API?

Yeah, it'll detect when user code incorrectly transfers ownership, either in (via `from_raw_fd`) or out (via `as_raw_fd`, either closing it manually in Rust, or more likely, IMO, returning it across an FFI boundary and calling close on the other side).

[Here's](https://android-review.googlesource.com/c/platform/frameworks/base/+/732305/2/core/jni/com_android_internal_content_NativeLibraryHelper.cpp) an example of a bug in Android that's the latter scenario: an fd is owned by a Java ParcelFIleDescriptor, and some C++ code grabbed it and passed it directly to some other code that expected to own the fd.

> Should `into_raw_fd` reset the tag to 0? I think it should since you're explicitly saying that the fd is no longer managed by Rust.

Oops, yes, I'll fix this shortly. Before merging this PR in general, I need to find somewhere to add some tests, but I'm not entirely clear on where they should live, since they're Android-only (and behavior will also depend on what version of Android it's running on). Any suggestions?

**Comment by Amanieu** [comment]
The documentation for `as_raw_fd` should be updated to say that behavior is undefined if the returned fd is closed even if `mem::forget` is used. It should mention Android's ownership tagging as an example and recommend the use of `into_raw_fd` instead.

The documentation for `from_raw_fd` should also state that the fd must not currently be owned by any other Rust fd wrapper (e.g. `File`). Again Android's fd tagging should be mentioned as an example of how this can cause issues in practice.

**Comment by pnkfelix** [comment]
[discussed](https://zulip-archive.rust-lang.org/238009tcompilermeetings/21643weeklymeeting2020080554818.html#206148208) at today's T-compiler triage meeting.

The T-compiler members present at the meeting didn't have any strong opinions here. We agreed with @nikomatsakis that this warrants an FCP of some sort.

There was general consensus that this is, at its heart, a question of what the publicly specified API for Rust's `File` should be; the actual implementation code (which T-compiler maintains) is pretty trivial.

So we collectively decided at the meeting that while this should require an FCP, it should really solely require an FCP of the members of T-libs; T-compiler is going to opt out of participating in the FCP.

Thus, I am removing T-compiler, adding T-libs, and will go see if I can find a T-libs member to start up the FCP process for this PR.

**Comment by Amanieu** [comment]
@rfcbot fcp merge

**Comment by jmgao** [comment]
> The documentation for `as_raw_fd` should be updated to say that behavior is undefined if the returned fd is closed even if `mem::forget` is used. It should mention Android's ownership tagging as an example and recommend the use of `into_raw_fd` instead.
> 
> The documentation for `from_raw_fd` should also state that the fd must not currently be owned by any other Rust fd wrapper (e.g. `File`). Again Android's fd tagging should be mentioned as an example of how this can cause issues in practice.

Somewhat of a nitpicky question: is "undefined behavior" the proper phrase to use for this? It's definitely incorrect, but it can't lead to the compiler using the fact that you did that to make assumptions that transitively breaks code elsewhere. The phrasing I'm thinking of now is something along the lines of:

`AsRawFd`:
```diff
 This method does not pass ownership of the raw file descriptor to the caller. The descriptor is only guaranteed to be valid while the original object has not yet been destroyed.
+Do not attempt to take ownership of this file descriptor (e.g. by using `FromRawFd`). On Android API level 30 and beyond, file descriptor ownership is enforced, which will lead to the process aborting if this occurs.
```

`FromRawFd`:
```diff
 This function consumes ownership of the specified file descriptor. The returned object will take responsibility for closing it when the object goes out of scope.
+The file descriptor consumed must not be already owned by another object. On Android API level 30 and beyond, file descriptor ownership is enforced, which will lead to the process aborting if this occurs.
```

**Comment by Amanieu** [comment]
Undefined behavior doesn't have to be linked to the compiler, it can come from the library as well: in this case it means that behavior may change in the future and may lead to unsoundness via unsafe assumptions (i.e. writing to the wrong FD).

I think that undefined behavior is appropriate in this case: Android happens to catch these cases, but other platforms do not.

**Comment by nikomatsakis** [comment]
Hey @KodrAus @SimonSapin @sfackler and @withoutboats -- any thoughts on the FCP here? (I'm trying to move things along for things in my review queue, but also this change seems like a genuine improvement.)

**Comment by KodrAus** [comment]
@rfcbot reviewed

**Comment by the8472** [comment]
> The documentation for `from_raw_fd` should also state that the fd must not currently be owned by any other Rust fd wrapper (e.g. `File`)

That would conflict with temporarily constructing a wrapper to use standard IO APIs from another wrapper type which doesn't support those APIs.

https://github.com/rust-lang/rust/blob/4e9c93b63eeed83b79eed17a8e180a2703a7d115/library/std/src/sys/unix/kernel_copy.rs#L421-L428

**Comment by nikomatsakis** [comment]
Perhaps the documentation just needs to be more precise?  I note that you use `ManuallyDrop` there, so maybe it's not so much that one cannot *construct* a `File` as allow it to be dropped?

**Comment by the8472** [comment]
If I read the proposed changes correctly then attempting to construct a temporary `ManuallyDrop<File>` from say `StdinLock` would try to retag an already tagged file descriptor which would then panic.

**Comment by jmgao** [comment]
Yes, that's correct. Maybe we could solve this with a `BorrowedFile` that has the same interface as a `File`, without maintaining ownership of the fd?

**Comment by the8472** [comment]
For this particular use-case, yes. But it prevents the more general pattern of building a `ManuallyDrop<T> where T: FromRawFd` from a `U: AsRawFd`. The issue is that any wrappers around a file descriptors doesn't expose all possible operations on a file descriptor. A similar case would be `std::fs::ReadDir` which also happens to be a handle to a directory which may be useful for other things, such as the family of `-at` family of syscalls, e.g. via the [openat crate](https://docs.rs/openat/0.1.19/openat/struct.Dir.html)

Perhaps `from_raw_fd` should be changed to only set a tag if none is present and otherweise store `tag: 0` in `FileDesc` and then panic on drop if it is 0. That would force the use of `ManuallyDrop` for such temporary conversions.

**Comment by the8472** [comment]
> `BorrowedFile`

If `AsRawFd` returned a descriptor with a lifetime then a `BorrowedFd<'a, T> where T: FromRawFd` and a `Deref` impl with `Target = T`  would be great way to express a temporary conversion that's guaranteed to have a shorter lifetime than the wrapper from which it was derived. Alas, `AsRawFd` returns `RawFd` without lifetime.

**Comment by the8472** [comment]
Could something like this work?

```rust
trait FromRawFd {
  fn from_wrapper(other: &'a impl AsRawFd) -> BorrowedFd<'a, Self> {
     BorrowedFd {inner: ManuallyDrop::new(Self::from_raw_fd(other.as_raw_fd())), phantom: PhantomData}
  }
}
```

**Comment by jmgao** [comment]
> Perhaps `from_raw_fd` should be changed to only set a tag if none is present and otherweise store `tag: 0` in `FileDesc` and then panic on drop if it is 0. That would force the use of `ManuallyDrop` for such temporary conversions.

I think this would be reasonable. Do we want to enshrine `ManuallyDrop<File>` as the right way to call those helpers on RawFd, though?

(It's a shame that RawFd is a typedef to int instead of being an opaque type that we could add methods onto, but I assume that ship has sailed.)

File: library/std/src/sys/unix/fd.rs [hunk 1]
```diff
@@ -8,9 +8,82 @@ use crate::sys_common::AsInner;
...
+mod android {
+    use crate::sync::atomic::{AtomicU64, Ordering};
+    use libc::c_int;
+
+    #[derive(Debug)]
+    pub struct CloseTag(u64);
+
+    static FD_TAG: AtomicU64 = AtomicU64::new(0);
+
+    fn next_tag() -> u64 {
```
Comment (the8472) [inline library/std/src/sys/unix/fd.rs:24]: This could use a comment that it will never return 0 which is good to know since 0 has special meaning.

**Comment by the8472** [comment]
> I think this would be reasonable. Do we want to enshrine `ManuallyDrop<File>` as the right way to call those helpers on RawFd, though?

There are many ways to use a RawFd. But if one wants to combine AsRawFd and FromRawFd then ManuallyDrop is a necessary but not sufficient part of the solution.

From the [current documentation of `as_raw_fd`](https://doc.rust-lang.org/std/os/unix/io/trait.AsRawFd.html#tymethod.as_raw_fd):

> This method does not pass ownership of the raw file descriptor to the caller. The descriptor is only guaranteed to be valid while the original object has not yet been destroyed.

Which means that whatever you use it for must not close the fd (as it does not pass ownership) and it must not use it longer than the lifetime of the original.

`ManuallyDrop` provides the 1st behavior but does not ensure the 2nd as it carries no lifetime bound. But that has always been a problem with `AsRawFd`. Hence the `from_wrapper` idea in my previous comment.

**Comment by nikomatsakis** [comment]
Hey @rust-lang/libs, I see that the FCP has completed, but I think [this concern](https://github.com/rust-lang/rust/pull/74860#issuecomment-727018292) raised by @the8472 merits an official response:

> That would conflict with temporarily constructing a wrapper to use standard IO APIs from another wrapper type which doesn't support those APIs.

My impression is that we don't have a workaround or alternative for this pattern, so I'm wondering whether you all consider this a blocking concern to be resolved.

**Comment by m-ou-se** [comment]
Thanks for the ping @nikomatsakis. I've added it to the agenda for the libs meeting, which starts in a few hours.

**Comment by m-ou-se** [comment]
A few of the places that use `ManuallyDrop<File>` or `ManuallyDrop<FileDesc>` or `ManuallyDrop<TcpStream>` (etc.) as a stand in for a 'borrowed file descriptor':

In std:
- https://github.com/rust-lang/rust/blob/7d747db0d5dd8f08f2efb073e2e77a34553465a7/library/std/src/sys/unix/stdio.rs#L17 (And five more in that file.)
- https://github.com/rust-lang/rust/blob/7d747db0d5dd8f08f2efb073e2e77a34553465a7/library/std/src/sys/unix/kernel_copy.rs#L423

In `wasmtime`:
- https://github.com/bytecodealliance/wasmtime/blob/3d606a01e50517de50adab7d1161c4057fb20aed/crates/wasi-common/src/sys/unix/mod.rs#L46-L51

In other crates:
- https://github.com/CleverCloud/amq-protocol/blob/a055e71bf51efd0f417308863bd5817fcdcae268/tcp/src/lib.rs#L64-L65
- https://github.com/siiptuo/pio/blob/28be2ae2857dc11b45a8292a81d84aa78f64f436/src/output.rs#L74
- https://github.com/YorickPeterse/inko/blob/267d09e38b1e5064b9c4638f1d0e639220e31c51/vm/src/closable.rs#L105

There's probably more, but it's a bit hard to search for.

**Comment by m-ou-se** [comment]
It seems to me that because of
https://github.com/rust-lang/rust/blob/7d747db0d5dd8f08f2efb073e2e77a34553465a7/library/std/src/sys/unix/stdio.rs#L36-L39
even a simple `println!("hello"); println!("world");` would already break, as it calls `write` twice, which would end up tagging stdout twice. Am I missing something?

**Comment by jmgao** [comment]
> My impression is that we don't have a workaround or alternative for this pattern, so I'm wondering whether you all consider this a blocking concern to be resolved.

IMO this is definitely a blocking concern. I think the behavior proposed by @the8472 might be the best option:

> Perhaps `from_raw_fd` should be changed to only set a tag if none is present and otherweise store `tag: 0` in `FileDesc` and then panic on drop if it is 0. That would force the use of `ManuallyDrop` for such temporary conversions.

**Comment by KodrAus** [comment]
@rfcbot cancel

**Comment by the8472** [comment]
This API is also new. Since CI is setup to only test against a fairly old minimum API level this change won't be exercised in CI and thus it won't catch broken uses.

**Comment by crlf0710** [comment]
@jmgao Ping from triage! What's the next steps here?

**Comment by crlf0710** [comment]
@jmgao Ping from triage, i'm closing this due to inactivity. Feel free to reopen or create a new PR when you have time to work on this again. Thanks!

**Pull request closed without merging**